test/e2e/providers/aws/aws.go

```diff
@@ -0,0 +1,387 @@
+			a.instanceType)
+	}
+	return nil, err
+}
+
+// getClusterWorkerSGID gets worker security group id from the existing cluster or returns an error.
+// This function is exposed for testing purpose.
+func (a *awsProvider) getClusterWorkerSGID(infraID string) (string, error) {
+	sg, err := a.ec2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: aws.StringSlice([]string{fmt.Sprintf("%s-worker-sg", infraID)}),
+			},
+			{
+				Name:   aws.String("tag:" + infraIDTagKeyPrefix + infraID),
+				Values: aws.StringSlice([]string{infraIDTagValue}),
+			},
+		},
+	})
+	if err != nil {
+		return "", err
+	}
+	if sg == nil || len(sg.SecurityGroups) < 1 {
+		return "", fmt.Errorf("no security group is found for the cluster worker nodes")
+	}
+	return *sg.SecurityGroups[0].GroupId, nil
+}
+
+// GetVPCByInfrastructure finds the VPC of an infrastructure and returns the VPC struct or an error.
+func (a *awsProvider) getVPCByInfrastructure(infraID string) (*ec2.Vpc, error) {
+	res, err := a.ec2.DescribeVpcs(&ec2.DescribeVpcsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:" + infraIDTagKeyPrefix + infraID),
+				Values: aws.StringSlice([]string{infraIDTagValue}),
+			},
+			{
+				Name:   aws.String("state"),
+				Values: aws.StringSlice([]string{"available"}),
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(res.Vpcs) < 1 {
+		return nil, fmt.Errorf("failed to find the VPC of the infrastructure")
+	} else if len(res.Vpcs) > 1 {
+		log.Printf("more than one VPC is found, using %s", *res.Vpcs[0].VpcId)
+	}
+	return res.Vpcs[0], nil
+}
+
+// getIAMWorkerRole gets worker IAM information from the existing cluster including IAM arn or an error.
+// This function is exposed for testing purpose.
+func (a *awsProvider) getIAMWorkerRole(infraID string) (*ec2.IamInstanceProfileSpecification, error) {
+	iamspc, err := a.iam.GetInstanceProfile(&iam.GetInstanceProfileInput{
+		InstanceProfileName: aws.String(fmt.Sprintf("%s-worker-profile", infraID)),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &ec2.IamInstanceProfileSpecification{
+		Arn: iamspc.InstanceProfile.Arn,
+	}, nil
+}
+
+// GenerateMachineSet generates the machineset object which is aws provider specific
+func (a *awsProvider) GenerateMachineSet(withWindowsLabel bool) (*mapi.MachineSet, error) {
+	clusterName, err := a.getInfraID()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get infrastructure id %v", err)
+	}
+
+	instanceProfile, err := a.getIAMWorkerRole(clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get instance profile %v", err)
+	}
+
+	sgID, err := a.getClusterWorkerSGID(clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get securoty group id: %v", err)
+	}
+
+	subnet, err := a.getSubnet(clusterName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get subnet: %v", err)
+	}
```

Comment:
For each cloud provider we will have to reimplement all of these functions. Have you considered finding a Linux worker MachineSet and gathering the required fields from there?
The benefits of that is that we will not need to make cloud provider api calls for each cloud provider. That will make it easier to add vSphere, Azure, and eventually GCP.

There should always be a linux worker MachineSet to base our MachineSet off of, or WMCO will not be able to run.

Reply:
> For each cloud provider we will have to reimplement all of these functions.

We don't have to. The only method that needs to be implemeted is `CreateMachineSet`.

> Have you considered finding a Linux worker MachineSet and gathering the required fields from there?

That is good point. @PratikMahajan and I did think of it but the main problem was fetching the  latest Windows AMI for which we have to make an explicit call . When we started doing it, we thought let's copy the code from WNI AWS provider but we exposed the single method that every cloud provider needs to implement which is `GenerateMachineSet`. In future, say if we are not able to figure a field and we would like to copy it from MachineSet we should still be able to do that.

> There should always be a linux worker MachineSet to base our MachineSet off of, or WMCO will not be able to run.

This is not entirely true, we can create it in a zone where there is no existing linux machineset. As of now, installer is creating those machinesets in the zones available but that doesn't mean you cannot specify the zones as part of installation.

Reply:
If it gets messy as we add more cloud providers we can re-evaluate, for now this is fine